### PR TITLE
schedulefree optimizers

### PR DESCRIPTION
**Pull request by winglian**
# What does this PR do?

integrates meta's https://github.com/facebookresearch/schedule_free for adamw & sgd

https://twitter.com/aaron_defazio/status/1776320004465582331

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@muellerzr  @younesbelkada  @pacman100

### CODE REVIEW

**Comment by muellerzr** [comment]
FYI this will need https://github.com/huggingface/accelerate/pull/2631 as we need to upstream accelerate's ability to call these things on a wrapped optimizer

**Comment by danielhanchen** [comment]
Some thoughts:

* I was trying to ask Aaron et al on Twitter if they did any transformer experiments, but to no avail. They said a paper will come in 1 or 2 months.
* Aaron et al's past work on D-Adaptation won a best ICML paper, with their follow up work being Prodigy - but both on transformers did similar or worse than AdamW. https://twitter.com/danielhanchen/status/1775547139248341125
* Superconvergence + LR range finder + Fast AI's Ranger21 optimizer was the goto optimizer for CNNs, and worked fabulously well, but on transformers, the learning rate range finder sadi 1e-3 was the best, whilst 1e-5 was better. However, the 1 cycle learning rate stuck. https://github.com/huggingface/transformers/issues/16013
* A huge issue is this needs tuning??! But how about a well tuned AdamW? Eg see https://twitter.com/kellerjordan0/status/1776716388037529843 which outperformed it using a tuned SGD.

I'm just a little bit reserved for now since the author themselves aren't providing any transformer benchmarks, nor have they compared their CNN baselines to superconvergence, which is the goto standard for fast training for CNNs. Likewise https://parameterfree.com/2023/08/30/yet-another-icml-award-fiasco/ wasn't pleasant.

**Comment by PhilipMay** [comment]
Should be very easy to test this on Phi-2 or TinyLlama when the implementation works?

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3117,6 +3145,9 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             `torch.Tensor`: The tensor with training loss on this batch.
         """
         model.train()
+        if "ScheduleFree" in self.optimizer.__class__.__name__:
```
Comment (younesbelkada) [inline src/transformers/trainer.py:3148]: maybe instead of checking the class name here we could inject an attribute `_hf_schedule_free_optim` to make sure we can support that in the future for other shcedule free optimizers, what do you think?

**Review by younesbelkada — COMMENTED** [review]
Great work @winglian ! 🤩 I left one minor comment, wdyt?

**Comment by PhilipMay** [comment]
This PR should maybe also add a few lines to the README about "how to use this".

**Comment by muellerzr** [comment]
We've merged the accelerate portion in, so if anyone is trying this out in distributed fashions, you can do `pip install git+https://github.com/huggingface/accelerate` :)

File: src/transformers/trainer.py (hunk 1 above)
Comment (winglian) [reply src/transformers/trainer.py:3148]: that would be on the Trainer class, right?

Comment (winglian) [reply src/transformers/trainer.py:3148]: so the place that makes the most sense to set that would be in `get_optimizer_cls_and_kwargs` but that is a `@staticmethod` so has no access to the trainer object. We could do something along the lines of 
```python
setattr(self.optimizer, "_hf_schedule_free_optim", True)
```
after we instantiate the `optimizer_cls` but we would still have to do some sort of class name detection.

Alternatively we could pass another value in the return tuple specific to schedule_free optimizers (but that feels worse)

**Comment by bratao** [comment]
There is any chance of this making into the main branch? I and other confirmed that the results are real. Thank you @winglian

**Review by pacman100 — COMMENTED** [review]
Super useful addition of scheduler free optimizers @winglian! It would be great to document the usage along with a minimal example.

**Comment by CoffeeVampir3** [comment]
Is their any remaining work I could contribute towards getting this PR merged? 

Cheers

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3117,6 +3145,9 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
+            self.optimizer.eval()
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/transformers/trainer.py]: thanks. fixed

**Comment by winglian** [comment]
@pacman100 @muellerzr @younesbelkada Can we get a new review to get this merged? Since the last check, I rebased, added some fixes and docs.

File: setup.py [hunk 3]
```diff
@@ -161,6 +161,7 @@
     "sacremoses",
     "safetensors>=0.4.1",
     "sagemaker>=2.31.0",
+    "schedulefree==1.2.1",
```
Comment (muellerzr) [inline setup.py:164]: I don't think we need to pin this aggressively.

**Review by muellerzr — APPROVED** [review]
Thanks! Overall LG2M, let's pin schedulefree as a `>=` however.

Can you also run the quality checks? Afterwords at least from my end looks good to merge.

**Comment by winglian** [comment]
@muellerzr ran the `make quality`/lint and also added a smoke test to the test suite for schedule free adam

**Review by muellerzr — APPROVED** [review]
Thanks a bunch! cc @LysandreJik for final review

**Review by younesbelkada — APPROVED** [review]
Thanks a lot !

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Main comment is about the `getattr` logic in `get_optimizer_cls_and_kwargs`

**Comment by winglian** [comment]
Will get back to this soon. Not stale 😅

**Comment by bratao** [comment]
@winglian please don´t let it die

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -1409,6 +1410,34 @@ def optimizer_hook(param):
...
+                additional_optim_kwargs = adam_kwargs
+            elif args.optim == OptimizerNames.SCHEDULE_FREE_SGD:
+                optimizer_cls = SGDScheduleFree
+            else:
+                raise ValueError("Invalid schedulefree optimizer")
+            additional_optim_kwargs["weight_decay"] = args.weight_decay
+            additional_optim_kwargs["warmup_steps"] = args.warmup_steps
+            additional_optim_kwargs.update(
+                {
+                    "weight_lr_power": float(getattr(torch, optim_args.get("weight_lr_power", 2.0))),
```
[the remark this answers is not in the document]
Comment (tmm1) [reply src/transformers/trainer.py:1436]:
```suggestion
                    "weight_lr_power": float(optim_args.get("weight_lr_power", 2.0)),
```

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -1409,6 +1410,34 @@ def optimizer_hook(param):
...
+                    "r": float(getattr(torch, optim_args.get("r", 0.0))),
```
[the remark this answers is not in the document]
Comment (tmm1) [reply src/transformers/trainer.py:1437]:
```suggestion
                    "r": float(optim_args.get("r", 0.0)),
```

**Comment by winglian** [comment]
thanks for the fixes @tmm1 !

**Comment by tmm1** [comment]
@amyeroberts I addressed your comments. LMK what else is required to push this through!

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -3371,6 +3403,9 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             `torch.Tensor`: The tensor with training loss on this batch.
         """
         model.train()
+        if _is_schedule_free_optimizer(self.optimizer):
+            self.optimizer.train()
```
[the remark this answers is not in the document]
Comment (tmm1) [reply src/transformers/trainer.py:3407]: Would it be preferable to check if the optimizer has the eval/train methods and call them, instead of checking the class name?

cc @adefazio

[the remark this answers is not in the document]
Comment (adefazio) [reply src/transformers/trainer.py:3407]: Yes, the way I do it in my code is by checking for the eval/train methods, i.e. duck typing.

[the remark this answers is not in the document]
Comment (tmm1) [reply src/transformers/trainer.py:3407]: is this approach acceptable?

```suggestion
        if hasattr(self.optimizer, 'train') and callable(getattr(self.optimizer, 'train')):
            self.optimizer.train()
```

[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/trainer.py:3407]: Yes, I think this would be preferable as it's more extensible to other optimizers, and then puts the responsibility for management on implementation / checking on the optimizer implementation, rather than us within trainer

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating - looks good to me! 

cc @muellerzr to confirm the conditional check on the `eval` and `train` methods are OK

**Comment by amyeroberts** [comment]
@winglian It looks like we'll need a bit more iteration on the `eval` and `train` method checks as it's causing failing tests atm. For the code quality tests, could you run `make fixup` and push the changes?

**Comment by tmm1** [comment]
> we'll need a bit more iteration on the `eval` and `train` method checks as it's causing failing tests atm

looks like we need to revisit https://github.com/huggingface/accelerate/pull/2631 which i've done in https://github.com/huggingface/accelerate/pull/3055



> could you run `make fixup` and push the changes

:+1:

**Comment by fizzAI** [comment]
lgtm!
for what it's worth, there was some asking when this PR was originally opened about whether schedulefree's optims were effective on transformers; i did some testing a while ago, and SF adamw had (albiet marginally) a better loss/acc landscape than normal adamw on at least sequence classification tuning, even after tuning hyperparams.
Screenshot_20240830-084839
(purple line is ScheduleFreeAdamW, red line is just AdamW)

**Comment by tmm1** [comment]
@amyeroberts can we kick off the tests again? does the build pick up the latest accelerate release automatically?

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -1488,6 +1489,34 @@ def optimizer_hook(param):
...
             optimizer_kwargs.update(adam_kwargs)
+        elif args.optim in [
+            OptimizerNames.SCHEDULE_FREE_ADAMW,
+            OptimizerNames.SCHEDULE_FREE_SGD,
+        ]:
+            if not is_schedulefree_available():
+                raise ImportError(
+                    "You need to install `schedulefree` in order to use schedulefree optimizers"
+                    " install it with `pip install schedulefree`"
+                )
```
Comment (muellerzr) [inline src/transformers/trainer.py:1500]: We also should have a check (after this one) for if the accelerate version is < 0.30.0 raise an error

**Review by muellerzr — APPROVED** [review]
Thanks! Just a request for one more additional check for users who don't have the right minimum accelerate version. cc @LysandreJik for final 🤗

File: src/transformers/trainer.py (hunk 7 above)
Comment (tmm1) [reply src/transformers/trainer.py:1500]:
```suggestion
            if not is_schedulefree_available():
                raise ImportError(
                    "You need to install `schedulefree` in order to use schedulefree optimizers"
                    " install it with `pip install schedulefree`"
                )
            if not is_accelerate_available("0.30.0"):
                raise ImportError("You need to have `accelerate>=0.30.0` to be able to use schedulefree optimizers")
```

**Review by LysandreJik — APPROVED** [review]
Thanks for the PR @winglian!

LGTM

**Comment by winglian** [comment]
> Thanks for the PR @winglian!

all the credit goes to @tmm1 for getting this over the line!
